libwacom/libwacom.c

```diff
@@ -283,15 +285,28 @@ get_device_info (const char            *path,
 	}
 
 	*name = g_strdup (g_udev_device_get_sysfs_attr (device, "name"));
-	/* Try getting the name from the parent if that fails */
-	if (*name == NULL) {
-		GUdevDevice *parent;
+	*fw = g_strdup (g_udev_device_get_sysfs_attr (device, "fw_name"));
+	/* Try getting the name/fw_name from the parent if that fails */
+	if (*name == NULL || *fw == NULL) {
+		GUdevDevice *parent = g_udev_device_get_parent (device);
+
+		while (parent && (*name == NULL || *fw == NULL)) {
+			GUdevDevice *tmp;
+
+			if (*name == NULL)
+				*name = g_strdup (g_udev_device_get_sysfs_attr (parent, "name"));
+			if (*fw == NULL)
+				*fw = g_strdup (g_udev_device_get_sysfs_attr (device, "fw_name"));
```

Comment:
This one is `parent`:

```
*fw = g_strdup (g_udev_device_get_sysfs_attr (parent, "fw_name"));
```

With this change (and adding a missing `S_IROTH` in the driver) libwacom is able to read the firmware name :tada: 

Now I need to parse it to drop the version (`HUION_T203_200720` -> `HUION_T203`) and make the match to the fw name optional for backwards compatibility.